### PR TITLE
Adding --check to b3sum

### PR DESCRIPTION
**Pull request by phayes**
This is for adding `--check` to the `b3sum` utility.

### CODE REVIEW

**Comment by phayes** [comment]
OK I think this is ready for review.

File: b3sum/src/main.rs [hunk 1]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
...
+        let hasher = hasher.clone();
+        let mut check_hash = Vec::<u8>::with_capacity(hash.len());
+        let check_reader = hash_file(&hasher, &os_filename, mmap_disabled)
+            .with_context(|| format!("filename: {}", filename))?;
+        check_reader
+            .take(hash.len() as u64)
+            .read_to_end(&mut check_hash)?;
+
+        // Check if hashes match.
+        // TODO: Should this be constant time?
```
Comment (phayes) [inline b3sum/src/main.rs]: @oconnor663 - I'd like your opinion of if we should be doing a constant-time check for equality here.

File: b3sum/src/main.rs [hunk 2]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
...
+        if check_hash == hash {
```
Comment (phayes) [inline b3sum/src/main.rs]: @oconnor663 , I'd like your opinion of if we should be doing a constant-time check for equality here.

Comment (oconnor663) [reply b3sum/src/main.rs]: I don't see any reason not to, and it might matter in some very obscure scenario. Let's stick to constant-time. We could even do a constant-time comparison on the hex bytes directly, and save ourselves the trouble of parsing them.

File: b3sum/src/main.rs [hunk 3]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
...
+    let hasher = blake3::Hasher::new();
+    let mut num_failed = 0;
+
+    let reader = std::io::BufReader::new(input);
+    for (line_num, line) in reader.lines().enumerate() {
+        // Read the line and trim it
+        let line: String = line.with_context(|| format!("line {}", line_num + 1))?;
+        let line = line.trim();
+
+        // Skip the line if it's a comment or empty
```
Comment (oconnor663) [inline b3sum/src/main.rs]: Do these conditions come up in practice? What does coreutils do with these cases?

File: b3sum/src/main.rs [hunk 4]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
...
+        // Get the hash value and parse the hex value
+        let hash: &str = line.next().expect("Unexpected missing value");
+        let hash = hex::decode(hash)
+            .with_context(|| format!("error decoding hex on line {}", line_num + 1))?;
+
+        // Get the filename
+        let filename: &str = line
+            .next()
+            .ok_or_else(|| Error::msg(format!("missing filename on line {}", line_num + 1)))?;
+        let os_filename: std::ffi::OsString = filename.into();
```
Comment (oconnor663) [inline b3sum/src/main.rs]: This raises an interesting point. Should we try to handle really weird filenames and escaping? It looks like `md5sum * | md5sum --check` can handle some very nasty stuff, including filenames with newlines or invalid unicode.

File: b3sum/src/main.rs [hunk 5]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
...
+
+        // Hash the file
+        let hasher = hasher.clone();
+        let mut check_hash = Vec::<u8>::with_capacity(hash.len());
+        let check_reader = hash_file(&hasher, &os_filename, mmap_disabled)
+            .with_context(|| format!("filename: {}", filename))?;
+        check_reader
+            .take(hash.len() as u64)
+            .read_to_end(&mut check_hash)?;
```
Comment (oconnor663) [inline b3sum/src/main.rs]: I'm a little skeptical of deriving the length from the input hashes. Say for example that the input hashes are just 1 byte long. That could lead to a situation where you think you're checking file integrity, but you're really not checking much. This gets more problematic if it's possible for some attacker to _control_ the length of the hashes you check.

On the flip side, checking hashes longer than 32 bytes doesn't add any security, but people might get confused and think that it does.

Maybe just have `--check` always assume the default length? Maybe even conflict with `--length`? (Incidentally it should probably also conflict with `--keyed` and `--derive-key`.)

File: b3sum/src/main.rs [hunk 6]
```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
     }
 }
 
+fn check_file(input: impl Read, mmap_disabled: bool) -> Result<u64> {
```
Comment (oconnor663) [inline b3sum/src/main.rs]: It's a little confusing that you've got `check_files` and `check_file`, but it's the latter that loops over all the filenames?

File: b3sum/tests/test.rs [hunk 7]
```diff
@@ -134,3 +134,16 @@ fn test_raw_with_multi_files_is_an_error() {
...
+    let in_file = tempfile::NamedTempFile::new().unwrap();
+    in_file.as_file().write_all(b"foo").unwrap();
+    in_file.as_file().flush().unwrap();
+
+    // Check the resuts by doing a round-trip via stdin / stdout
+    let result = cmd!(b3sum_exe(), in_file.path())
+        .pipe(cmd!(b3sum_exe(), "--check"))
+        .run();
+    assert!(result.is_ok());
+}
```
Comment (oconnor663) [inline b3sum/tests/test.rs]: We'll definitely want to test multiple inputs. Also whatever we do with weird whitespace and invalid unicode, we should try to test here.

File: b3sum/src/main.rs (hunk 2 above)
Comment (sneves) [reply b3sum/src/main.rs]: You would have to lowercase both strings in that case, which is about the same trouble as parsing the hex string into bytes and comparing there.

File: b3sum/src/main.rs (hunk 5 above)
Comment (phayes) [reply b3sum/src/main.rs]: I was trying to ensure that all output from `b3sum` is also valid input for `b3sum --check`.    So that if you do `b3sum * --length=128` and pass that back into `b3sum --check` it should work. 

I would recommend two things:

1. Throw a warning if the hash lengths are under some threshold at which we consider the security compromised.  We should do this for both checking and generating the checksum file. 

2. Add a "strict" mode where warnings are upgraded to errors. 

This has some precedence.  `b2sum-rs` has the following options that we could emulate:
```
The following five options are useful only when verifying checksums:
      --ignore-missing  don't fail or report status for missing files
      --quiet           don't print OK for each successfully verified file
      --status         don't output anything, status code shows success
      --strict          exit non-zero for improperly formatted checksum lines
``` 

However, if we do decide to error on checking non-32-byte checksums, then we should make sure we at least generate a warning when generating them.

File: b3sum/src/main.rs (hunk 4 above)
Comment (phayes) [reply b3sum/src/main.rs]: I'm a big fan of drawing the line at non-unicode filenames.     Non-unicode filenames need to be relegated to the dustbin of history  :)  

But if you really think we need to support non-unicode filenames, then I can work on adding support. 

I agree that we should add tests for all sorts of pathological (unicode) filename cases though.

File: b3sum/src/main.rs (hunk 2 above)
Comment (oconnor663) [reply b3sum/src/main.rs]: Are we allowed to assume that the output we're parsing came from `b3sum`? If so, maybe we don't have to support uppercase hex if we don't emit any? But anyway, parsing hex to bytes is also fine with me.

File: b3sum/src/main.rs (hunk 3 above)
Comment (phayes) [reply b3sum/src/main.rs]: I was following along with the behavior here: 

https://github.com/jtdowney/b2sum-rs/blob/master/src/main.rs#L127-L129

Comment (phayes) [reply b3sum/src/main.rs]: `shasum` (on my mac at least) allows one trailing empty line, but errors on more than one consecutive empty line.

File: b3sum/src/main.rs (hunk 6 above)
Comment (phayes) [reply b3sum/src/main.rs]: Correct.  I will rename things to make it clearer.

File: b3sum/src/main.rs (hunk 4 above)
Comment (oconnor663) [reply b3sum/src/main.rs]: On the one hand, it might not be worth the effort. On the other hand, if we treat everything as raw bytes, maybe we get this "support" for free? (But then again, Windows paths aren't amenable to raw bytes, are they? Not making Window support an annoyance might be a good reason to stick with Unicode/UTF-8.)

Comment (phayes) [reply b3sum/src/main.rs]: Also, deciding what is whitespace and a line-end when interpreting raw bytes is a PITA.    I think I would want to bring in `regex::bytes` as a dependency to handle this if we decide to support non-unicode.

File: b3sum/src/main.rs (hunk 3 above)
Comment (oconnor663) [reply b3sum/src/main.rs]: If we can simplify things by demanding that the input format is exactly the same as the tool output, I'm all for simplifying.

File: b3sum/src/main.rs (hunk 5 above)
Comment (oconnor663) [reply b3sum/src/main.rs]: I think if we have `check` throw an error saying "this only works with 32-byte hashes", we'll get 90% of the value with 10% of the effort. Trying to determine "security compromised" in general is too vague, and it depends on the application. (Collision-resistant checksums need to be 256 bits, but MACs can usually get away with 128, for example.)

**Comment by oconnor663** [comment]
I've pushed a different `--check` implementation to master, and documented it at https://github.com/BLAKE3-team/BLAKE3/blob/master/b3sum/what_does_check_do.md. If you get a chance, and let me know what you think at https://github.com/BLAKE3-team/BLAKE3/issues/33?
